File: frontend/src/api/agents/copyInjection.ts
/**
 * Copy Injection Agent API
 */

import { api } from '@/api'

export interface TemplateSummary {
  id: string
  name: string
}

export interface CopyInjectionInput {
  template_id: string
  raw_copy: string
  product_name?: string
  product_category?: string
}

export interface PlacementSummary {
  placeholder: string
  content_preview: string
  filled: boolean
}

export interface CopyInjectionOutput {
  html: string
  placeholders_found: string[]
  placements: PlacementSummary[]
  images_generated: number
  success: boolean
  error_message?: string
}

export interface ExtractPlaceholdersInput {
  html_template: string
}

export interface ExtractPlaceholdersOutput {
  placeholders: string[]
  count: number
}

export const copyInjectionApi = {
  /**
   * Fetch all available templates for the dropdown
   */
  async getTemplates(): Promise<TemplateSummary[]> {
    return api.get<TemplateSummary[]>('/agents/copyinjection/templates')
  },


  /**
   * Process template and copy, returns JSON response with metadata
   */
  async process(input: CopyInjectionInput): Promise<CopyInjectionOutput> {
    return api.post<CopyInjectionOutput>('/agents/copyinjection', input)
  },
}

File: backend/app/api/v1/routers/copy_injection.py
"""
Copy Injection Agent Router
============================

API endpoints for Agent 1 - Copy & Image Injection.
"""

import logging

from fastapi import APIRouter
from pydantic import BaseModel, Field

from app.services.agents.copy_injection import (
    CopyInjectionAgent,
    CopyInjectionInput,
    CopyInjectionOutput,
    TemplateService,
    TemplateSummary,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/copyinjection", tags=["agents"])


# ---------------------------------------------------------------------------
# Templates
# ---------------------------------------------------------------------------

@router.get("/templates", response_model=list[TemplateSummary])
async def get_templates() -> list[TemplateSummary]:
    """Return all available templates for the frontend dropdown."""
    logger.info("GET /copyinjection/templates")
    return TemplateService.get_all()




# ---------------------------------------------------------------------------
# Legacy placeholder extraction (kept for debugging)
# ---------------------------------------------------------------------------

class ExtractPlaceholdersInput(BaseModel):
    """Input for placeholder extraction."""

    html_template: str = Field(..., description="HTML template to extract placeholders from")


class ExtractPlaceholdersOutput(BaseModel):
    """Output from placeholder extraction."""

    placeholders: list[str] = Field(default_factory=list)
    count: int = Field(default=0)


@router.post("/extract-placeholders", response_model=ExtractPlaceholdersOutput)
async def extract_placeholders(payload: ExtractPlaceholdersInput) -> ExtractPlaceholdersOutput:
    """Extract all placeholders from an HTML template (debug utility)."""
    logger.info(
        "POST /copyinjection/extract-placeholders: template_length=%d",
        len(payload.html_template),
    )
    agent = CopyInjectionAgent()
    placeholders = agent.extract_placeholders(payload.html_template)
    return ExtractPlaceholdersOutput(placeholders=placeholders, count=len(placeholders))


# ---------------------------------------------------------------------------
# Main injection endpoint
# ---------------------------------------------------------------------------

@router.post("", response_model=CopyInjectionOutput)
async def inject_copy(payload: CopyInjectionInput) -> CopyInjectionOutput:
    """
    Fill a template with marketing copy and return the generated HTML + metadata.
    Preview and ZIP packaging are handled entirely client-side.
    """
    logger.info(
        "POST /copyinjection: template_id=%s, copy_length=%d",
        payload.template_id,
        len(payload.raw_copy),
    )
    agent = CopyInjectionAgent()
    return await agent.process(payload)

